keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,495 @@
+
+No.
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+- No to existing API objects that doesn't use this feature.
+- For API objects that use this feature:
+  - API type: Pod
+  - Estimated increase in size: 
+    - New field `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution`. Which is a `NodeSelector` struct.
+    - NodeSelector:
+      - NodeSelectorTerms (slice header): 24 bytes
+      - Total: 24 bytes
+    - NodeSelectorTerm:
+      - MatchExpressions (slice header): 24 bytes
+      - MatchFields (slice header): 24 bytes
+      - Total: 48 bytes
+    - NodeSelectorRequirement:
+      - Key (string): 16 bytes (assuming an average size)
+      - Operator (string): 16 bytes (assuming an average size)
+      - Values (slice header): 24 bytes (assuming one element)
+      - Values[0] (string): 16 bytes (assuming an average size)
+      - Total: 88 bytes
+    - The estimated size of the `requiredDuringSchedulingRequiredDuringExecution` would be approximately 176 bytes(assuming we only defined one rule in it).
+    
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+If `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` field is defined, this feature will slightly increase the pod scheduling time and increase the possibility of pods being evicted.
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+The scheduler have to process `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` parameter which may result in some small increase in CPU usage.
+A new controller will be added to kube-controller-manager and will also result in some small increase in CPU usage.
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+No.
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+During the downtime of API server and/or etcd:
+
+- Running workloads that don't need to be evicted function well.
+- Running workloads that need to be evicted by unmatched `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` will stay in current state
+as API requests will be rejected.
+
+###### What are other known failure modes?
+
+N/A
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+- 2023-12-13: Initial draft KEP
+
+## Drawbacks
+
+N/A
+
+## Alternatives
+
+- taints
+  
+  Taint can achieve this feature in some scenarios. But nodeaffinity has more operators for users to choose from to cover more scenarios.
+  Also taints are meant to protect nodes, so they don't respect PDB.
+
+- descheduler
+
```

Comment:
I'm guessing that you could enable the API server feature gate (only) and use the descheduler for the descheduling part, assuming we update that out-of-tree code.

Does this sound right?

Reply:
Yes, we can do that. However this requires some code changes in descheduler.